File: script/make/protocol/js/NoticeProtocol.js
export default class NoticeProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 50001: {

                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 50001: {
                // 公告列表
                const data = [];
                let dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--dataLength >= 0) {
                    // 
                    // 公告ID
                    const dataDataNoticeId = view.getBigUint64(offset, false);
                    offset = offset + 8;
                    // 收到时间
                    const dataDataReceiveTime = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 读取时间
                    const dataDataReadTime = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 标题
                    const dataDataTitleLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const dataDataTitleArray = new Uint8Array(view.buffer.slice(offset, offset + dataDataTitleLength));
                    const dataDataTitle = textDecoder.decode(dataDataTitleArray);
                    offset = offset + dataDataTitleLength;
                    // 内容
                    const dataDataContentLength = view.getUint16(offset, false);
                    offset = offset + 2;
                    const dataDataContentArray = new Uint8Array(view.buffer.slice(offset, offset + dataDataContentLength));
                    const dataDataContent = textDecoder.decode(dataDataContentArray);
                    offset = offset + dataDataContentLength;
                    // object
                    const dataData = {"noticeId": dataDataNoticeId, "receiveTime": dataDataReceiveTime, "readTime": dataDataReadTime, "title": dataDataTitle, "content": dataDataContent};
                    // add
                    data.push(dataData);
                }
                return data;
            }
            case 50002: {
                // 
                // 范围
                const dataScope = view.getUint8(offset, false);
                offset = offset + 1;
                // 类型
                const dataType = view.getUint8(offset, false);
                offset = offset + 1;
                // 标题
                const dataTitleLength = view.getUint16(offset, false);
                offset = offset + 2;
                const dataTitleArray = new Uint8Array(view.buffer.slice(offset, offset + dataTitleLength));
                const dataTitle = textDecoder.decode(dataTitleArray);
                offset = offset + dataTitleLength;
                // 消息
                const dataMsgLength = view.getUint16(offset, false);
                offset = offset + 2;
                const dataMsgArray = new Uint8Array(view.buffer.slice(offset, offset + dataMsgLength));
                const dataMsg = textDecoder.decode(dataMsgArray);
                offset = offset + dataMsgLength;
                // object
                const data = {"scope": dataScope, "type": dataType, "title": dataTitle, "msg": dataMsg};
                return data;
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/erl/notice_protocol.erl
-module(notice_protocol).
-export([decode/2, encode/2]).
-include("notice.hrl").

-spec decode(Protocol :: non_neg_integer(), Binary :: binary()) -> {ok, [integer() | binary() | list()]} | {error, Protocol :: non_neg_integer(), Binary :: binary()}.
decode(50001, _Rest_ = <<_/binary>>) ->
    {ok, {}};


decode(Protocol, Binary) ->
    {error, Protocol, Binary}.


-spec encode(Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()) -> {ok, binary()} | {error, Protocol :: non_neg_integer(), Data :: atom() | tuple() | binary() | list()}.
encode(50001, Data) ->
    Data50001 = <<(encode_data_50001(<<>>, 0, Data))/binary>>,
    {ok, <<(byte_size(Data50001)):16, 50001:16, Data50001/binary>>};

encode(50002, {Scope, Type, Title, Msg}) ->
    Data50002 = <<Scope:8, Type:8, (byte_size(Title)):16, (Title)/binary, (byte_size(Msg)):16, (Msg)/binary>>,
    {ok, <<(byte_size(Data50002)):16, 50002:16, Data50002/binary>>};

encode(Protocol, Data) ->
    {error, Protocol, Data}.

encode_data_50001(Acc = <<_/binary>>, Length, []) ->
    <<Length:16, Acc/binary>>;
encode_data_50001(Acc = <<_/binary>>, Length, [#notice_role{notice_id = NoticeId, receive_time = ReceiveTime, read_time = ReadTime, title = Title, content = Content} | Data]) ->
    encode_data_50001(<<Acc/binary, NoticeId:64, ReceiveTime:32, ReadTime:32, (byte_size(Title)):16, (Title)/binary, (byte_size(Content)):16, (Content)/binary>>, Length + 1, Data).

